src/app-layer-enip.c

```diff
@@ -0,0 +1,743 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL
+ * THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \file
+ *
+ * \author Kevin Wong <kwong@solananetworks.com>
+ *
+ * App-layer parser for ENIP protocol
+ *
+ */
+
+#include "suricata-common.h"
+
+#include "util-debug.h"
+#include "util-byte.h"
+#include "util-enum.h"
+#include "util-mem.h"
+#include "util-misc.h"
+
+#include "stream.h"
+
+#include "app-layer-protos.h"
+#include "app-layer-parser.h"
+#include "app-layer-enip.h"
+#include "app-layer-enip-common.h"
+
+#include "app-layer-detect-proto.h"
+
+#include "conf.h"
+#include "decode.h"
+
+
+
+void *ENIPGetTx(void *alstate, uint64_t tx_id)
+{
+    ENIPState *enip_state = (ENIPState *)alstate;
+    ENIPTransaction *tx = NULL;
+
+    /* fast track: try the current tx */
+    if (enip_state->curr && enip_state->curr->tx_num == tx_id + 1)
+        return enip_state->curr;
+
+    /* fast track:
+     * if the prev tx_id is equal to the stored tx ptr, we can
+     * use this shortcut to get to the next. */
+    if (enip_state->iter) {
+        if (tx_id == enip_state->iter->tx_num) {
+            tx = TAILQ_NEXT(enip_state->iter, next);
+            if (tx && tx->tx_num == tx_id + 1) {
+                enip_state->iter = tx;
+                return tx;
+            }
+        }
+    }
+
+    /* no luck with the fast tracks, do the full list walk */
+    TAILQ_FOREACH(tx, &enip_state->tx_list, next) {
+        SCLogDebug("tx->tx_num %u, tx_id %"PRIu64, tx->tx_num, (tx_id+1));
+        if ((tx_id+1) != tx->tx_num)
+            continue;
+
+        SCLogDebug("returning tx %p", tx);
+        enip_state->iter = tx;
+        return tx;
+    }
+
+    return NULL;
+}
+
+uint64_t ENIPGetTxCnt(void *alstate)
+{
+    ENIPState *enip_state = (ENIPState *)alstate;
+    return (uint64_t)enip_state->transaction_max;
+}
+
+
+int ENIPGetAlstateProgress(void *tx, uint8_t direction)
```

Comment:
This function always returns 1, indicating the tx is complete (as 1 is returned in ENIPGetAlstateProgressCompletionStatus). This means that the tx is marked complete on just a request.

There are several ways to handle this better. The http parser for example, tracks the state of each direction of a tx separately. The DNS parser has a simpler approach: for toserver to be done, use 1. For toclient to be done, use 2. See https://github.com/inliniac/suricata/commit/c419f33f441882ba5c080337667db5591df9eb2c